File: nodejs/src/client/workflows.ts
/**
 * Workflows API: list, get (with nodes), create, get_graph, deploy, get_writer.
 * Backend: workflows microservice (/workflows/workflows, graph, projects/:id/deploy).
 *
 * A workflow is a DAG of nodes (ingestion → transformation → export). This
 * namespace is the single surface for the workflow entity and its lifecycle;
 * the former `flows` namespace has been folded in here.
 *
 * `get_writer` is a low-level stream-writer escape hatch — see `@internal` note.
 * Live event writes use the Loxtep stream data plane, not HTTP POST.
 */

import type { RStreamsSdk } from '../rstreams/leo-runtime.js';
import type { LoxtepHttpClient } from '../http/client.js';
import type {
  WorkflowsListFilters,
  WorkflowsListResponse,
  GetWorkflowGraphResponse,
  DeployInput,
  DeployResponse,
  CreateWorkflowInput,
  SaveWorkflowBundleInput,
  SaveWorkflowBundleResult,
  SaveWorkflowBundleResponse,
} from './workflows-types.js';
import type {
  Flow,
  FlowWithNodes,
  FlowNode,
  FlowWriter,
  GetWriterOptions,
} from './flow-types.js';
import type { DefinitionValidationErrorEntry } from '../errors/types.js';
import { DefinitionValidationError } from '../errors/validation.js';
import { StreamingError } from '../errors/streaming.js';
import { createQueueWriter } from '../rstreams/event-bridge.js';
import type { WriteOptions } from './queue-types.js';
import { resolveIngestionQueueName } from './flow-queue-resolve.js';

const WORKFLOWS_BASE = '/workflows/workflows';
const PROJECTS_BASE = '/workflows/projects';

export interface WorkflowsApiDeps {
  /** Stream runtime; required for get_writer().close() */
  rsdk?: RStreamsSdk;
  /** Lazy stream runtime resolver (attempts observe.stream_config() when rsdk is not set). */
  get_rsdk?: () => Promise<RStreamsSdk | undefined>;
}

function buildQueryString(params: Record<string, string | number | boolean | undefined>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null) continue;
    search.set(k, String(v));
  }
  const qs = search.toString();
  return qs ? `?${qs}` : '';
}

function validateEventAgainstDefinition(
  event: unknown,
  definition: Record<string, unknown>
): DefinitionValidationErrorEntry[] {
  const errors: DefinitionValidationErrorEntry[] = [];
  const required = definition.required as string[] | undefined;
  if (
    Array.isArray(required) &&
    required.length > 0 &&
    event !== null &&
    typeof event === 'object'
  ) {
    const obj = event as Record<string, unknown>;
    for (const key of required) {
      if (!(key in obj) || obj[key] === undefined) {
        errors.push({ path: key, message: `Missing required field: ${key}` });
      }
    }
  }
  return errors;
}

/**
 * Determine whether an error is transient and should be retried.
 * Transient: network errors, timeouts, throttling (429), 5xx responses.
 * Non-transient: 4xx (except 429), auth failures — fail immediately.
 */
export function isTransientError(err: unknown): boolean {
  if (err === null || err === undefined) return false;

  // Check for status code on the error object (HTTP response errors)
  const statusCode =
    typeof (err as { status_code?: unknown }).status_code === 'number'
      ? (err as { status_code: number }).status_code
      : typeof (err as { statusCode?: unknown }).statusCode === 'number'
        ? (err as { statusCode: number }).statusCode
        : typeof (err as { status?: unknown }).status === 'number'
          ? (err as { status: number }).status
          : undefined;

  if (statusCode !== undefined) {
    // 429 (throttling) is transient
    if (statusCode === 429) return true;
    // 5xx responses are transient
    if (statusCode >= 500 && statusCode < 600) return true;
    // Other 4xx are non-transient (auth failures, validation errors, etc.)
    if (statusCode >= 400 && statusCode < 500) return false;
  }

  // Check error code strings
  const code =
    typeof (err as { code?: unknown }).code === 'string'
      ? (err as { code: string }).code
      : undefined;

  if (code) {
    const transientCodes = new Set([
      'ECONNRESET',
      'ECONNREFUSED',
      'ENOTFOUND',
      'ETIMEDOUT',
      'EPIPE',
      'EAI_AGAIN',
      'RATE_LIMIT_EXCEEDED',
      'RATE_LIMIT_ERROR',
      'SERVICE_UNAVAILABLE',
      'GATEWAY_TIMEOUT',
      'INTERNAL_SERVER_ERROR',
    ]);
    if (transientCodes.has(code)) return true;

    // Auth-related codes are non-transient
    const nonTransientCodes = new Set([
      'AUTHENTICATION_ERROR',
      'AUTHORIZATION_ERROR',
      'VALIDATION_ERROR',
      'NOT_FOUND',
    ]);
    if (nonTransientCodes.has(code)) return false;
  }

  // Check error message for network-related keywords
  const message =
    typeof (err as { message?: unknown }).message === 'string'
      ? (err as { message: string }).message.toLowerCase()
      : '';

  if (
    message.includes('timeout') ||
    message.includes('econnreset') ||
    message.includes('econnrefused') ||
    message.includes('network') ||
    message.includes('socket hang up')
  ) {
    return true;
  }

  // Default: treat unknown errors as transient to allow retry
  return true;
}

/**
 * Sleep for the given number of milliseconds. Extracted for testability.
 */
export function sleep(ms: number): Promise<void> {
  if (ms <= 0) return Promise.resolve();
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export type WorkflowsApi = {
  list: (filters: WorkflowsListFilters) => Promise<WorkflowsListResponse['data']>;
  get: (id: string) => Promise<FlowWithNodes>;
  create: (input: CreateWorkflowInput) => Promise<Flow>;
  get_graph: (
    workflow_id: string,
    project_id: string
  ) => Promise<GetWorkflowGraphResponse['data']>;
  deploy: (input: DeployInput) => Promise<DeployResponse['data']>;
  save_workflow_bundle: (
    project_id: string,
    input: SaveWorkflowBundleInput
  ) => Promise<SaveWorkflowBundleResult>;
  /**
   * @internal
   * Low-level stream-writer escape hatch. Requires explicit `bot_id` and an
   * output queue (or `environment_prefix` + `project_id` to resolve it).
   * Undocumented for customers — use `data_products.get_writer(name)` instead.
   */
  get_writer: (workflow_id: string, options?: GetWriterOptions) => Promise<FlowWriter>;
};

export function createWorkflowsApi(
  http: LoxtepHttpClient,
  deps?: WorkflowsApiDeps
): WorkflowsApi {
  const api: WorkflowsApi = {
    async list(filters: WorkflowsListFilters): Promise<WorkflowsListResponse['data']> {
      const params: Record<string, string | number | undefined> = {
        project_id: filters.project_id,
        page: filters.page ?? 1,
        page_size: filters.page_size ?? 100,
      };
      if (filters.status) params.status = filters.status;
      if (filters.search) params.search = filters.search;
      const qs = buildQueryString(params);
      const res = await http.get<WorkflowsListResponse>(`${WORKFLOWS_BASE}${qs}`);
      return res.data;
    },

    async get(id: string): Promise<FlowWithNodes> {
      const flowRes = await http.get<{ success: true; data: Flow }>(
        `${WORKFLOWS_BASE}/${encodeURIComponent(id)}`
      );
      const flow = flowRes.data;
      let nodes: FlowNode[] = [];
      try {
        const nodesRes = await http.get<{ success: true; data: { items: FlowNode[] } }>(
          `${WORKFLOWS_BASE}/${encodeURIComponent(id)}/nodes`
        );
        const payload = (nodesRes as { data?: { items?: FlowNode[] } }).data ?? nodesRes;
        nodes = (payload as { items?: FlowNode[] }).items ?? [];
      } catch {
        // Nodes endpoint may not exist or may 404; workflow still has node_count
      }
      return { ...flow, nodes };
    },

    async create(input: CreateWorkflowInput): Promise<Flow> {
      const res = await http.post<{ success: true; data: Flow }>(WORKFLOWS_BASE, input);
      return res.data;
    },

    async get_graph(
      workflow_id: string,
      project_id: string
    ): Promise<GetWorkflowGraphResponse['data']> {
      const qs = buildQueryString({ project_id });
      const res = await http.get<GetWorkflowGraphResponse>(
        `${WORKFLOWS_BASE}/${encodeURIComponent(workflow_id)}/graph${qs}`
      );
      return res.data;
    },

    async deploy(input: DeployInput): Promise<DeployResponse['data']> {
      const { project_id, instance_id, version_id, force_redeploy } = input;
      const res = await http.post<DeployResponse>(
        `${PROJECTS_BASE}/${encodeURIComponent(project_id)}/deploy`,
        { instance_id, version_id, force_redeploy: force_redeploy ?? false }
      );
      return res.data;
    },

    async save_workflow_bundle(
      project_id: string,
      input: SaveWorkflowBundleInput
    ): Promise<SaveWorkflowBundleResult> {
      const res = await http.post<SaveWorkflowBundleResponse>(
        `${PROJECTS_BASE}/${encodeURIComponent(project_id)}/workflow-bundle`,
        {
          files: input.files,
          dry_run: input.dry_run ?? false,
        }
      );
      return res.data;
    },

    async get_writer(workflow_id: string, options?: GetWriterOptions): Promise<FlowWriter> {
      // Resolve the stream runtime, source bot, and destination queue up front; the rstreams `load`
      // stream (created inside createQueueWriter) owns buffering, batching, backoff, and checkpointing.
      const rsdk = deps?.rsdk ?? (await deps?.get_rsdk?.());
      if (!rsdk) {
        throw new StreamingError(
          'Stream bus configuration missing. Set LEO_* environment variables or add `streams` to ~/.loxtep/config.json. ' +
            'Pass `streams` in LoxtepClientOptions, or ensure your instance stream environment is configured.',
          {
            details: {
              workflow_id,
              hint: 'Run `loxtep init` to configure stream bus, or set LEO_* env vars from your instance.',
            },
          }
        );
      }

      const botId = options?.bot_id;
      if (!botId) {
        throw new StreamingError(
          'workflows.get_writer requires bot_id in options (stream writer identity on the bus).',
          { details: { workflow_id } }
        );
      }

      let queueName = options?.output_queue_name;
      if (!queueName && options?.environment_prefix && options?.project_id) {
        const flowRes = await http.get<{ success: true; data: Flow } | Flow>(
          `${WORKFLOWS_BASE}/${encodeURIComponent(workflow_id)}`
        );
        const flowData =
          flowRes && typeof flowRes === 'object' && 'data' in flowRes
            ? (flowRes as { data: Flow }).data
            : (flowRes as Flow);
        let nodes: FlowNode[] = [];
        try {
          const nodesRes = await http.get<{ success: true; data: { items: FlowNode[] } }>(
            `${WORKFLOWS_BASE}/${encodeURIComponent(workflow_id)}/nodes`
          );
          const payload = (nodesRes as { data?: { items?: FlowNode[] } }).data ?? nodesRes;
          nodes = (payload as { items?: FlowNode[] }).items ?? [];
        } catch {
          /* ignore */
        }
        const flowWithNodes: FlowWithNodes = { ...flowData, nodes };
        queueName = resolveIngestionQueueName(flowWithNodes, options.environment_prefix);
      }
      if (!queueName) {
        throw new StreamingError(
          'workflows.get_writer requires output_queue_name, or environment_prefix + project_id so the SDK can resolve the ingestion queue from the workflow.',
          { details: { workflow_id } }
        );
      }

      const writer = createQueueWriter(
        rsdk,
        botId,
        queueName,
        () =>
          new StreamingError(
            'Cannot write to a closed writer. Create a new writer via workflows.get_writer().',
            { details: { workflow_id } }
          )
      );

      // Optional per-event definition validation wraps the passthrough writer.
      const validate = options?.validate_definition === true && !!options?.definition;
      if (!validate) return writer;
      const onError = options?.on_validation_error ?? 'reject';
      return {
        write(event: unknown, writeOptions?: WriteOptions): void {
          if (options?.definition) {
            const errors = validateEventAgainstDefinition(event, options.definition);
            if (errors.length > 0) {
              if (onError === 'reject') {
                throw new DefinitionValidationError(
                  'Event failed definition validation',
                  options.definition_version ?? '',
                  errors
                );
              }
              if (onError === 'warn') {
                console.warn('[workflows writer] Definition validation failed:', errors);
              }
              if (onError === 'skip') return;
            }
          }
          writer.write(event, writeOptions);
        },
        close: () => writer.close(),
      };
    },
  };
  return api;
}

File: python/src/loxtep/workflows.py
"""
Workflows API: list, get (with nodes), create, get_graph, deploy, get_writer.
Backend: workflows microservice (/workflows/workflows, graph, projects/:id/deploy).
snake_case per backend conventions.

The former ``flows`` namespace has been folded in here (same backend entity).
``get_writer`` is a low-level stream-writer escape hatch — internal; customers
should use ``data_products.get_writer``.
"""

from typing import Any, Literal, Optional
from urllib.parse import quote

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient

WORKFLOWS_BASE = "/workflows/workflows"
PROJECTS_BASE = "/workflows/projects"

WorkflowType = Literal["ingestion", "enrichment", "delivery"]
"""Required by backend POST /workflows (nodejs/src/client/flow-types.ts FlowCreateInput)."""


def _query_string(params: dict[str, Any]) -> str:
    parts = [f"{k}={quote(str(v))}" for k, v in params.items() if v is not None]
    return "?" + "&".join(parts) if parts else ""


def _data(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


class WorkflowsApi:
    """Sync client for workflow list, get, create, graph, deploy, and writer."""

    def __init__(self, http: LoxtepHttpClient, stream_config: Optional[Any] = None) -> None:
        self._http = http
        self._stream_config = stream_config

    def list(
        self,
        project_id: str,
        *,
        page: int = 1,
        page_size: int = 100,
        status: Optional[str] = None,
        search: Optional[str] = None,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {
            "project_id": project_id,
            "page": page,
            "page_size": page_size,
        }
        if status is not None:
            params["status"] = status
        if search is not None:
            params["search"] = search
        qs = _query_string(params)
        res = self._http.get(f"{WORKFLOWS_BASE}{qs}")
        return _data(res)

    def get(self, id: str) -> dict[str, Any]:
        flow_res = self._http.get(f"{WORKFLOWS_BASE}/{quote(id)}")
        flow = _data(flow_res)
        nodes: list[dict[str, Any]] = []
        try:
            nodes_res = self._http.get(f"{WORKFLOWS_BASE}/{quote(id)}/nodes")
            data = _data(nodes_res)
            nodes = data.get("items", []) if isinstance(data, dict) else []
        except Exception:
            pass
        out: dict[str, Any] = {"nodes": nodes}
        if isinstance(flow, dict):
            out["workflow"] = flow
        return out

    def create(
        self,
        name: str,
        project_id: str,
        *,
        workflow_type: WorkflowType,
        domain_id: str,
        connection_id: Optional[str] = None,
        template_id: Optional[str] = None,
        description: Optional[str] = None,
        configuration: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Create a workflow. `workflow_type` and `domain_id` are required by the
        backend (`POST /workflows`) — omitting them 500s with a raw DB error."""
        body: dict[str, Any] = {
            "name": name,
            "project_id": project_id,
            "workflow_type": workflow_type,
            "domain_id": domain_id,
        }
        if connection_id is not None:
            body["connection_id"] = connection_id
        if template_id is not None:
            body["template_id"] = template_id
        if description is not None:
            body["description"] = description
        if configuration is not None:
            body["configuration"] = configuration
        res = self._http.post(WORKFLOWS_BASE, body)
        return _data(res)

    def get_graph(self, workflow_id: str, project_id: str) -> dict[str, Any]:
        qs = _query_string({"project_id": project_id})
        path = f"{WORKFLOWS_BASE}/{quote(workflow_id)}/graph{qs}"
        res = self._http.get(path)
        return _data(res)

    def deploy(
        self,
        project_id: str,
        instance_id: str,
        *,
        version_id: Optional[str] = None,
        force_redeploy: bool = False,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "instance_id": instance_id,
            "force_redeploy": force_redeploy,
        }
        if version_id is not None:
            body["version_id"] = version_id
        path = f"{PROJECTS_BASE}/{quote(project_id)}/deploy"
        res = self._http.post(path, body)
        return _data(res)

    def get_writer(
        self, workflow_id: str, *, bot_id: Optional[str] = None, queue_name: Optional[str] = None
    ) -> Any:
        """Low-level stream-writer escape hatch. Internal — prefer
        ``data_products.get_writer``.

        Uses the Kinesis stream bus when the client has stream config and a
        queue is available; otherwise an HTTP writer.
        """
        cfg = self._stream_config
        if cfg is not None and getattr(cfg, "is_writable", False) and queue_name:
            from .rstreams import LoxtepStreamWriter

            return LoxtepStreamWriter(cfg, bot_id or f"sdk-writer-{workflow_id}", queue_name)
        return WorkflowWriter(workflow_id=workflow_id, http=self._http)


class WorkflowWriter:
    """Sync workflow writer: write(event), close(). Internal escape hatch."""

    def __init__(self, workflow_id: str, http: LoxtepHttpClient) -> None:
        self._workflow_id = workflow_id
        self._http = http

    def write(self, event: dict[str, Any]) -> None:
        self._http.post(f"{WORKFLOWS_BASE}/{self._workflow_id}/events", event)

    def close(self) -> None:
        pass


class AsyncWorkflowsApi:
    """Async client for workflow list, get, create, graph, deploy, and writer."""

    def __init__(self, http: AsyncLoxtepHttpClient, stream_config: Optional[Any] = None) -> None:
        self._http = http
        self._stream_config = stream_config

    async def list(
        self,
        project_id: str,
        *,
        page: int = 1,
        page_size: int = 100,
        status: Optional[str] = None,
        search: Optional[str] = None,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {
            "project_id": project_id,
            "page": page,
            "page_size": page_size,
        }
        if status is not None:
            params["status"] = status
        if search is not None:
            params["search"] = search
        qs = _query_string(params)
        res = await self._http.get(f"{WORKFLOWS_BASE}{qs}")
        return _data(res)

    async def get(self, id: str) -> dict[str, Any]:
        flow_res = await self._http.get(f"{WORKFLOWS_BASE}/{quote(id)}")
        flow = _data(flow_res)
        nodes: list[dict[str, Any]] = []
        try:
            nodes_res = await self._http.get(f"{WORKFLOWS_BASE}/{quote(id)}/nodes")
            data = _data(nodes_res)
            nodes = data.get("items", []) if isinstance(data, dict) else []
        except Exception:
            pass
        out: dict[str, Any] = {"nodes": nodes}
        if isinstance(flow, dict):
            out["workflow"] = flow
        return out

    async def create(
        self,
        name: str,
        project_id: str,
        *,
        workflow_type: WorkflowType,
        domain_id: str,
        connection_id: Optional[str] = None,
        template_id: Optional[str] = None,
        description: Optional[str] = None,
        configuration: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Create a workflow. `workflow_type` and `domain_id` are required by the
        backend (`POST /workflows`) — omitting them 500s with a raw DB error."""
        body: dict[str, Any] = {
            "name": name,
            "project_id": project_id,
            "workflow_type": workflow_type,
            "domain_id": domain_id,
        }
        if connection_id is not None:
            body["connection_id"] = connection_id
        if template_id is not None:
            body["template_id"] = template_id
        if description is not None:
            body["description"] = description
        if configuration is not None:
            body["configuration"] = configuration
        res = await self._http.post(WORKFLOWS_BASE, body)
        return _data(res)

    async def get_graph(self, workflow_id: str, project_id: str) -> dict[str, Any]:
        qs = _query_string({"project_id": project_id})
        path = f"{WORKFLOWS_BASE}/{quote(workflow_id)}/graph{qs}"
        res = await self._http.get(path)
        return _data(res)

    async def deploy(
        self,
        project_id: str,
        instance_id: str,
        *,
        version_id: Optional[str] = None,
        force_redeploy: bool = False,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "instance_id": instance_id,
            "force_redeploy": force_redeploy,
        }
        if version_id is not None:
            body["version_id"] = version_id
        path = f"{PROJECTS_BASE}/{quote(project_id)}/deploy"
        res = await self._http.post(path, body)
        return _data(res)

    def get_writer(
        self, workflow_id: str, *, bot_id: Optional[str] = None, queue_name: Optional[str] = None
    ) -> Any:
        """Low-level stream-writer escape hatch. Internal — prefer
        ``data_products.get_writer``. Uses the Kinesis bus when configured."""
        cfg = self._stream_config
        if cfg is not None and getattr(cfg, "is_writable", False) and queue_name:
            from .rstreams import AsyncLoxtepStreamWriter

            return AsyncLoxtepStreamWriter(cfg, bot_id or f"sdk-writer-{workflow_id}", queue_name)
        return AsyncWorkflowWriter(workflow_id=workflow_id, http=self._http)


class AsyncWorkflowWriter:
    """Async workflow writer. Internal escape hatch."""

    def __init__(self, workflow_id: str, http: AsyncLoxtepHttpClient) -> None:
        self._workflow_id = workflow_id
        self._http = http

    async def write(self, event: dict[str, Any]) -> None:
        await self._http.post(f"{WORKFLOWS_BASE}/{self._workflow_id}/events", event)

    async def close(self) -> None:
        pass
